src/mlpack/core/data/load_categorical_csv.hpp

```diff
@@ -0,0 +1,342 @@
+    if (MapPolicy::NeedsFirstPass)
+    {
+      // In this case we must pass everything we parse to the MapPolicy.
+      size_t dim = 0;
+      std::stringstream lineStream;
+      std::string token;
+
+      lineStream.clear();
+      lineStream.str(line);
+
+      while (lineStream.good())
+      {
+        std::getline(lineStream, token, delim);
+        // Remove whitespace from either side
+        trim(token);
+
+        if (token[0] == '"' && token[token.size() - 1] != '"')
+        {
+          std::string tok = token;
+
+          while (token[token.size() - 1] != '"')
+          {
+            tok += delim;
+            std::getline(lineStream, token, delim);
+            tok += token;
+          }
+          token = tok;
+        }
+        info.template MapFirstPass<T>(std::move(token), dim++);
+      }
+    }
+  }
+}
+
+template<typename T, typename MapPolicy>
+void LoadCSV::InitializeMapper(size_t& rows, size_t& cols, DatasetMapper<MapPolicy>& info)
+{
+  // Take a pass through the file.  If the DatasetMapper policy requires it,
+  // we will pass everything string through MapString().  This might be useful
+  // if, e.g., the MapPolicy needs to find which dimensions are numeric or
+  // categorical.
+
+  // Reset to the start of the file.
+  inFile.clear();
+  inFile.seekg(0, std::ios::beg);
+  rows = 0;
+  cols = 0;
+
+  // First, count the number of rows in the file (this is the dimensionality).
+  std::string line;
+  while (std::getline(inFile, line))
+    ++rows;
+
+  // Reset the DatasetInfo object, if needed.
+  if (info.Dimensionality() == 0)
+  {
+    info.SetDimensionality(rows);
+  }
+  else if (info.Dimensionality() != rows)
+  {
+    std::ostringstream oss;
+    oss << "data::LoadCSV(): given DatasetInfo has dimensionality "
+    << info.Dimensionality() << ", but data has dimensionality "
+    << rows;
+    throw std::invalid_argument(oss.str());
+  }
+
+  // Now, jump back to the beginning of the file.
+  inFile.clear();
+  inFile.seekg(0, std::ios::beg);
+  rows = 0;
+
+  while (std::getline(inFile, line))
+  {
+    ++rows;
+    // Remove whitespaces from either side
+    trim(line);
+    if (rows == 1)
+    {
+      // Extract the number of columns.
+      std::pair<size_t, size_t> dimen = GetMatrixSize<false>(inFile, delim);
+      cols = dimen.second;
+    }
+
+    // I guess this is technically a second pass, but that's ok... still the
+    // same idea...
+    if (MapPolicy::NeedsFirstPass)
+    {
+      std::string str(line.begin(), line.end());
+      std::stringstream lineStream;
+      std::string token;
+
+      lineStream.clear();
+      lineStream.str(line);
+
+      while (lineStream.good())
+      {
+        std::getline(lineStream, token, delim);
+        // Remove whitespace from either side
+        trim(token);
+
+        if (token[0] == '"' && token[token.size() - 1] != '"')
+        {
+          std::string tok = token;
+          while (token[token.size() - 1] != '"')
```

Comment:
You need to check the indentation here, it should be two spaces

Reply:
It is two spaces, I counted the spaces.
It is showing 2 spaces in my editor.